engine/resources/hetzner_vm.go

```diff
@@ -0,0 +1,804 @@
+		return false, errwrap.Wrapf(err, "waitFor(rescueModeOK) exited early")
+	}
+	return checkOK, nil
+}
+
+// getServerUpdate pings the Hetzner API for up-to-date server info.
+func (obj *HetznerVMRes) getServerUpdate(ctx context.Context) error {
+	if obj.init.Debug {
+		obj.init.Logf("getServerUpdate()")
+	}
+	server, _, err := obj.client.Server.GetByName(ctx, obj.Name())
+	if err != nil {
+		return errwrap.Wrapf(err, "failed serverupdate request")
+	}
+	obj.server = server
+	return nil
+}
+
+// Cmp compares two HetznerVMRes structs. Returns nil if the comparison holds
+// true, and otherwise an error is thrown to identify the difference.
+func (obj *HetznerVMRes) Cmp(r engine.Res) error {
+	// check if empty
+	if obj == nil && r == nil {
+		return nil
+	}
+	if (obj == nil) != (r == nil) {
+		return fmt.Errorf("one resource is empty")
+	}
+	// compare types
+	res, ok := r.(*HetznerVMRes)
+	if !ok {
+		return fmt.Errorf("not a %s", obj.Kind())
+	}
+	// compare resource fields
+	if obj.State != res.State {
+		return fmt.Errorf("state differs")
+	}
+	if obj.APIToken != res.APIToken {
+		return fmt.Errorf("apitoken differs")
+	}
+	if obj.ServerType != res.ServerType {
+		return fmt.Errorf("servertype differs")
+	}
+	if obj.Datacenter != res.Datacenter {
+		return fmt.Errorf("datacenter differs")
+	}
+	if obj.Image != res.Image {
+		return fmt.Errorf("image differs")
+	}
+	if obj.ServerRescueMode != res.ServerRescueMode {
+		return fmt.Errorf("serverrescuemode differs")
+	}
+	for i, key := range obj.ServerRescueSSHKeys {
+		if key != res.ServerRescueSSHKeys[i] {
+			return fmt.Errorf("serverrescuekeys differs")
+		}
+	}
+	return nil
+}
+
+// cmpServerSpecs compares the server specifications between the virtual
+// HetznerVMRes resource and its corresponding Hetzner instance. Returns true if
+// ServerType, Datacenter and Image match. Returns an error is the server is
+// unavailable (`absent`).
+func (obj HetznerVMRes) cmpServerSpecs() (bool, error) {
+	if obj.init.Debug {
+		obj.init.Logf("cmpServerSpecs()")
+	}
+	if obj.server == nil {
+		return false, fmt.Errorf("server is unavailable")
+	}
+	if obj.ServerType != obj.server.ServerType.Name {
+		return false, nil
+	}
+	if obj.Datacenter != obj.server.Datacenter.Name {
+		return false, nil
+	}
+	if obj.Image != obj.server.Image.Name {
+		return false, nil
+	}
+	return true, nil
+}
+
+// createServer checks if the servername is not already in use, builds the
+// serverconfig in hcloud-go format from resource params, requests a server
+// creation with that configuration, and waits until the creation is confirmed.
+// Errors occur when the server exists already, the client fails, or the wait
+// step exits early (context cancellation, client failure or timeout).
+// TODO: add createServer permission field
+func (obj *HetznerVMRes) createServer(ctx context.Context) error {
+	if obj.init.Debug {
+		obj.init.Logf("createServer()")
+	}
+	if obj.server != nil {
```

Comment:
are you sure obj.server is always up-to-date before createServer, deleteServer, and so on are called? Do we have a calling convention to ensure this? Should this happen in this function? Does it happen just once at resource start? What if we poll for ten minutes... Maybe things changed by then? LMK if this is handled correctly somewhere or not.

Reply:
I looked into this earlier, and I believe it is implemented correctly, but I agree that it might be good to have a more consistent design pattern/rule for it, or at least add some documentation on my current design somewhere. As it is now, CheckApply opens with a serverupdate request, and then requests a new update any time a change is made. Those updates happen inside the functions that push those changes though (after the operation, before returning, to validate the outcome). It's not super clear when reading through CheckApply, but it's there.

What would you prefer to make this more clear? Not sure where to write this kind of information. For now I will just add a note to the CheckApply description.